HeterogeneousCore/SonicTriton/src/TritonService.cc

```diff
@@ -0,0 +1,224 @@
+
+  const auto& modelServers = mit->servers;
+
+  if (!preferred.empty()) {
+    auto sit = modelServers.find(preferred);
+    //todo: add a "strict" parameter to stop execution if preferred server isn't found?
+    if (sit == modelServers.end())
+      edm::LogWarning("PreferredServer") << "Preferred server " << preferred << " for model " << model
+                                         << " not available, will choose another server";
+    else
+      return std::make_pair(findServer(preferred)->url, false);
+  }
+
+  //todo: use some algorithm to select server rather than just picking arbitrarily
+  auto sit = findServer(*modelServers.begin());
+  bool isFallbackCPU = sit->isFallback and !fallbackOpts_.useGPU;
+  return std::make_pair(sit->url, isFallbackCPU);
+}
+
+void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::ProcessContext const&) {
+  //only need fallback if there are unserved models
+  if (!fallbackOpts_.enable or unservedModels_.empty())
+    return;
+
+  std::string msg;
+  if (verbose_)
+    msg = "List of models for fallback server: ";
+  //all unserved models are provided by fallback server
+  auto sit = findServer(Server::fallbackName);
+  for (const auto& model : unservedModels_) {
+    auto mit = models_.insert(model).first;
+    mit->servers.insert(Server::fallbackName);
+    sit->models.insert(mit->name);
+    if (verbose_)
+      msg += mit->name + ", ";
+  }
+  if (verbose_)
+    edm::LogInfo("TritonService") << msg;
+
+  //assemble server start command
+  std::string command("triton");
+  if (fallbackOpts_.verbose)
+    command += " -v";
+  if (fallbackOpts_.useDocker)
+    command += " -d";
+  if (fallbackOpts_.useGPU)
+    command += " -g";
+  if (!fallbackOpts_.instanceName.empty())
+    command += " -n " + fallbackOpts_.instanceName;
+  if (fallbackOpts_.retries >= 0)
+    command += " -r " + std::to_string(fallbackOpts_.retries);
+  if (fallbackOpts_.wait >= 0)
+    command += " -w " + std::to_string(fallbackOpts_.wait);
+  for (const auto& model : unservedModels_) {
+    command += " -m " + model.path;
+  }
+  //don't need this anymore
+  unservedModels_.clear();
+
+  //get a random temporary directory if none specified
+  if (fallbackOpts_.tempDir.empty()) {
+    auto tmp_dir_path{std::filesystem::temp_directory_path() /= std::tmpnam(nullptr)};
+    fallbackOpts_.tempDir = tmp_dir_path.string();
+  }
+  //special case ".": use script default (temp dir = .$instanceName)
+  if (fallbackOpts_.tempDir != ".")
+    command += " -t " + fallbackOpts_.tempDir;
+
+  command += " start";
+
+  if (verbose_)
+    edm::LogInfo("TritonService") << command;
+  else
+    command += " >>/dev/null 2>>/dev/null";
+
+  //mark as started before executing in case of ctrl+c while command is running
+  startedFallback_ = true;
+  int rv = std::system(command.c_str());
+  if (rv != 0)
+    throw cms::Exception("FallbackFailed") << "Starting the fallback server failed with exit code " << rv;
```

Comment:
In the current setup, enabling the `verbose` option in the server will display the output of the `std::system()` calls (while the output is not displayed if `verbose` is disabled, which is the default). I'm actually not entirely sure what controls that behavior.

Is there a preferred way to handle this kind of system call? Maybe using popen/pclose?

Reply:
> I'm actually not entirely sure what controls that behavior.

Hah, now I remember that I did it: https://github.com/cms-sw/cmssw/blob/5c7b0a1053011c4e0686c35160a036987d066456/HeterogeneousCore/SonicTriton/src/TritonService.cc#L161-L164

Reply:
If the fallback server will never be used in production or in IB/PR tests, requiring the job to be re-run with `verbose` enabled could be acceptable (from the infrastructure point of view at least).

> Is there a preferred way to handle this kind of system call? Maybe using popen/pclose?

@Dr15Jones Would you have a suggestion?

Reply:
Nevermind, I missed the code being updated already.

Reply:
@dan131riley any suggestions?

Reply:
Just to be clear, I've since reimplemented this using popen/pclose. That was the only generic/supported approach that I found.

Reply:
`popen` gives you the advantage of being able to control the output, instead of having it mixed in more or less at random.  The downside is that the parent has to read the pipe; if it doesn't, the pipe can fill up which can stall the child process.